posthog/api/plugin.py

```diff
@@ -124,26 +127,89 @@ def destroy(self, request: request.Request, *args, **kwargs) -> Response:
+        for attachment in attachments:
+            new_plugin_config[attachment.key] = {
+                "uid": attachment.id,
+                "saved": True,
+                "size": attachment.file_size,
+                "name": attachment.file_name,
+                "type": attachment.content_type,
+            }
+        return new_plugin_config
+
     def create(self, validated_data: Dict, *args: Any, **kwargs: Any) -> PluginConfig:
         if not can_configure_plugins_via_api():
             raise ValidationError("Plugin configuration via the web is disabled!")
         request = self.context["request"]
         validated_data["team"] = request.user.team
+        self._fix_formdata_config_json(validated_data)
         plugin_config = super().create(validated_data)
+        self._update_plugin_attachments(plugin_config)
         reload_plugins_on_workers()
         return plugin_config
 
     def update(self, plugin_config: PluginConfig, validated_data: Dict, *args: Any, **kwargs: Any) -> PluginConfig:  # type: ignore
+        self._fix_formdata_config_json(validated_data)
         validated_data.pop("plugin", None)
         response = super().update(plugin_config, validated_data)
+        self._update_plugin_attachments(plugin_config)
         reload_plugins_on_workers()
         return response
 
+    # sending files via a multipart form puts the config JSON in a un-serialized format
+    def _fix_formdata_config_json(self, validated_data: dict):
+        request = self.context["request"]
+        if not validated_data.get("config", None) and request.POST.get("config", None):
+            validated_data["config"] = json.loads(request.POST["config"])
+
+    def _update_plugin_attachments(self, plugin_config: PluginConfig):
+        request = self.context["request"]
+        for key, file in request.FILES.items():
+            match = re.match(r"^add_attachment\[([^]]+)\]$", key)
+            if match:
+                self._update_plugin_attachment(plugin_config, match.group(1), file)
+        for key, file in request.POST.items():
+            match = re.match(r"^remove_attachment\[([^]]+)\]$", key)
+            if match:
+                self._update_plugin_attachment(plugin_config, match.group(1), None)
+
+    def _update_plugin_attachment(self, plugin_config: PluginConfig, key: str, file: Optional[UploadedFile]):
+        try:
+            plugin_attachment = PluginAttachment.objects.get(
+                team=plugin_config.team, plugin_config=plugin_config, key=key
+            )
+            if file:
+                plugin_attachment.content_type = file.content_type
+                plugin_attachment.file_name = file.name
+                plugin_attachment.file_size = file.size
+                plugin_attachment.contents = file.file.read()
+                plugin_attachment.save()
+            else:
+                plugin_attachment.delete()
+        except ObjectDoesNotExist:
```

Comment:
I think we can get rid of this!

```
plugin_attachment, created = PluginAttachment.objects.get_or_create(team=plugin_config.team, plugin_config=plugin_config, key=key)
```

From https://docs.djangoproject.com/en/dev/ref/models/querysets/#get-or-create

I believe it will also perform a django upsert under the hood.

Reply:
The thing is I don't always need to create the object. Sometimes I also want to delete it... and a "create" directly followed by "delete" is IMO worse practice than catching a `ObjectDoesNotExist`